### PR TITLE
All Goals Reachable setting

### PR DESCRIPTION
**Pull request by fenhl**
This implements #1196 as a variant of Beatable Only (which is now called Required Only in the GUI) by checking if all goal items are still reachable upon placing each item. Some things to note:

* I've only done minimal testing, i.e. checked that a seed generates successfully.
* This does not move “no logic” into this setting as proposed by @puddydoo. (I think it makes more sense to discuss that separately.)
* This does not yet consider Bingo goals. There is significant opposition to the idea of integrating AGR with Bingo. Maybe in a separate PR.

### CODE REVIEW

**Comment by LostThePirate** [comment]
Not a fan of the tooltip; I don't think the "speedrun category" analogy is appropriate here, especially since the *vast* majority of players do not come from a speedrunning background. I think these comparisons only muddy the explanation of the new categories, not clarify them.

**Comment by fenhl** [comment]
I wasn't aware of that. Removed that part.

**Comment by LostThePirate** [comment]
One of the most *vocal* segments of the community is the racers, who often also have a speedrunning background. But there are thousands more seeds rolled a day on the website than there are races, and we kinda have to assume that many of those are more casual players who don't come around the Discord for racing and whatnot.

**Comment by fenhl** [comment]
Question regarding Bingosync: Are there detailed rules anywhere? For example, I'm wondering what the definition of an “Adult Dungeon” is.

**Comment by LostThePirate** [comment]
Discord user DobbyDigital#1186 added the Bingo Hints system, so they could probably clarify further, but I believe they are generally based on/modified from vanilla OoT Bingo goals. Iirc, "dungeon" is specifically blue-warp dungeons, so Adult Dungeons would be the Temples.

**Comment by dragonbane0** [comment]
>  @dragonbane0 any ideas on how to fix this?

This usually happens when a tooltip line is too long. Try breaking earlier. Generally keep lines relatively short since there usually is more height then there is width.

**Comment by puddydoo** [comment]
I think "100%" still helps explain ALR without a speedrunning context.

**Comment by fenhl** [comment]
Since there is significant opposition to the idea of integrating AGR with Bingo, I will not be including it in this PR.

**Comment by fenhl** [comment]
> This usually happens when a tooltip line is too long. Try breaking earlier. Generally keep lines relatively short since there usually is more height then there is width.

That didn't fix it, not even after putting each word on its own line.

File: Location.py [hunk 1]
```diff
@@ -67,13 +67,17 @@ def set_rule(self, lambda_rule):
         self.access_rules = [lambda_rule]
 
 
-    def can_fill(self, state, item, check_access=True):
+    def can_fill(self, state, item, check_access=True, extra_location_checks=[]):
```
Comment (Zannick) [inline Location.py:70]:
```suggestion
    def can_fill(self, state, item, check_access=True, extra_location_checks=()):
```
Avoid mutable defaults.

File: Location.py [hunk 2]
```diff
@@ -67,13 +67,17 @@ def set_rule(self, lambda_rule):
...
         if self.minor_only and item.majoritem:
             return False
-        return (
-            not self.is_disabled() and 
-            self.can_fill_fast(item) and
-            (not check_access or state.search.spot_access(self, 'either')))
+        if self.is_disabled() or not self.can_fill_fast(item) or (check_access and not state.search.spot_access(self, 'either')):
+            return False
+        if len(extra_location_checks) == 0:
```
Comment (Zannick) [inline Location.py:75]:
```suggestion
        if not extra_location_checks:
```

File: Location.py [hunk 3]
```diff
@@ -67,13 +67,17 @@ def set_rule(self, lambda_rule):
...
+            return True
+        search_with_this = state.search.copy()
+        search_with_this.collect(item)
+        search_with_this.collect_locations(search_with_this.progression_locations() + extra_location_checks)
+        return all(search_with_this.visited(extra_location) for extra_location in extra_location_checks)
```
Comment (Zannick) [inline Location.py:80]:
```suggestion
        search_with_this.collect_locations(chain(search_with_this.progression_locations(), extra_location_checks))
        return all(map(search_with_this.visited, extra_location_checks))
```
You'll need `from itertools import chain` at the top if you take this suggestion.

**Review by Zannick — CHANGES_REQUESTED** [review]

**Comment by fenhl** [comment]
Suggestions applied.

**Review by Zannick — APPROVED** [review]

**Pull request merged**